pyo3-derive-backend/src/py_class.rs

```diff
@@ -133,25 +172,44 @@ fn impl_class(
         quote! {0}
     };
 
-    quote! {
-        impl ::pyo3::typeob::PyTypeInfo for #cls {
-            type Type = #cls;
+    // Create a variant of our generics with lifetime 'a prepended.
+    let mut gen_with_a = generics.clone();
+    gen_with_a.params.insert(0, parse_quote! { 'a });
+    let impl_with_a = gen_with_a.split_for_impl().0;
+
+    // Generate one PyTypeInfo per generic variant.
+    use quote::ToTokens;
+    let variant_iter: Box<dyn Iterator<Item = (String, TokenStream)>> = match attrs.variants {
```

Comment:
This part might get simpler by collecting in to a Vec.

Reply:
Will do!